### PR TITLE
Socket fixes for proxy

### PR DESCRIPTION
**Pull request by aznair**
Mainly fixing shutdown & destroy sequence, other minor bug fixes thrown in.
Proxy seems to work well(ish) with Chrome, not so much with Firefox.

@ercanucan please take a look at changes to proxy and python wrapper
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49174380%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49174413%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175347%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175392%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175500%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175586%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23issuecomment-169958701%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23issuecomment-170469687%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23issuecomment-169958701%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22proxy%20related%20bits%20lgtm%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A37%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7704593%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ercanucan%22%7D%7D%2C%20%7B%22body%22%3A%20%22%28Please%20remember%20to%20squash%20commits%20before%20merging%29%22%2C%20%22created_at%22%3A%20%222016-01-11T08%3A51%3A07Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20f611f3319c90c2fa046568e664a629fe315f2f0f%20endhost/scion_socket.py%2037%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175500%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22we%20can%20now%20remove%20this%20TODO%20now%20that%20we%20are%20calling%20deleteSCIONSocket%28%29%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A30%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7704593%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ercanucan%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_socket.py%3AL195-217%22%7D%2C%20%22Pull%20f611f3319c90c2fa046568e664a629fe315f2f0f%20endhost/scion_socket.py%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175347%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22let%27s%20return%200%20here%3F%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A28%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7704593%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ercanucan%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_socket.py%3AL127-136%22%7D%2C%20%22Pull%20f611f3319c90c2fa046568e664a629fe315f2f0f%20endhost/scion_proxy.py%2050%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49174380%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27m%20not%20sure%20this%20is%20necessary%20-%20the%20%60finally%3A%20cleanup%28%29%60%20calls%20will%20already%20take%20care%20of%20this.%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A15%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_proxy.py%3AL278-333%22%7D%2C%20%22Pull%20f611f3319c90c2fa046568e664a629fe315f2f0f%20endhost/scion_proxy.py%2063%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49174413%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Not%20strictly%20necessary%2C%20as%20it%27s%20already%20implied%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A16%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yeah%20I%20just%20like%20it%20better%20when%20things%20are%20explicit%2C%20but%20agree%20it%27s%20not%20necessary%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A31%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/801826%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/aznair%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_proxy.py%3AL278-333%22%7D%2C%20%22Pull%20f611f3319c90c2fa046568e664a629fe315f2f0f%20endhost/scion_socket.py%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/582%23discussion_r49175392%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22..and%20here%20as%20well%20%28return%200%29%22%2C%20%22created_at%22%3A%20%222016-01-08T10%3A28%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/7704593%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/ercanucan%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20endhost/scion_socket.py%3AL127-136%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull f611f3319c90c2fa046568e664a629fe315f2f0f endhost/scion_proxy.py 50'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#discussion_r49174380'>File: endhost/scion_proxy.py:L278-333</a></b>
- <a href='https://github.com/kormat'></a> I'm not sure this is necessary - the `finally: cleanup()` calls will already take care of this.
- [ ] <a href='#crh-comment-Pull f611f3319c90c2fa046568e664a629fe315f2f0f endhost/scion_proxy.py 63'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#discussion_r49174413'>File: endhost/scion_proxy.py:L278-333</a></b>
- <a href='https://github.com/kormat'></a> Not strictly necessary, as it's already implied
- <a href='https://github.com/aznair'></a> Yeah I just like it better when things are explicit, but agree it's not necessary
- [ ] <a href='#crh-comment-Pull f611f3319c90c2fa046568e664a629fe315f2f0f endhost/scion_socket.py 6'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#discussion_r49175347'>File: endhost/scion_socket.py:L127-136</a></b>
- <a href='https://github.com/ercanucan'></a> let's return 0 here?
- [ ] <a href='#crh-comment-Pull f611f3319c90c2fa046568e664a629fe315f2f0f endhost/scion_socket.py 8'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#discussion_r49175392'>File: endhost/scion_socket.py:L127-136</a></b>
- <a href='https://github.com/ercanucan'></a> ..and here as well (return 0)
- [ ] <a href='#crh-comment-Pull f611f3319c90c2fa046568e664a629fe315f2f0f endhost/scion_socket.py 37'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#discussion_r49175500'>File: endhost/scion_socket.py:L195-217</a></b>
- <a href='https://github.com/ercanucan'></a> we can now remove this TODO now that we are calling deleteSCIONSocket()
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/582#issuecomment-169958701'>General Comment</a></b>
- <a href='https://github.com/ercanucan'></a> proxy related bits lgtm
- <a href='https://github.com/kormat'></a> (Please remember to squash commits before merging)

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/582?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/582?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/582'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: endhost/scion_proxy.py [hunk 1]
```diff
@@ -278,50 +278,55 @@ def _read_write(self, target_sock):
...
             data = self._read()
-            if data:
-                self._write(data)
-            else:
-                self.shutdown.set()
+            if not data:
+                break
+            if not self._write(data):
+                break
+        self.rsock.close()
```
Comment (kormat) [inline endhost/scion_proxy.py]: I'm not sure this is necessary - the `finally: cleanup()` calls will already take care of this.

File: endhost/scion_proxy.py [hunk 2]
```diff
@@ -278,50 +278,55 @@ def _read_write(self, target_sock):
...
+            pass
         logging.debug("Done")
 
     def _read(self):
         try:
             return self.rsock.recv(BUFLEN)
         except OSError as e:
-            self.shutdown.set()
             logging.debug("Rsock closed: %s", e)
+            return None
```
Comment (kormat) [inline endhost/scion_proxy.py]: Not strictly necessary, as it's already implied

File: endhost/scion_socket.py [hunk 3]
```diff
@@ -127,6 +127,9 @@ def send(self, msg):
         :returns: The number of bytes sent.
         :rtype: int
         """
+        if self.fd == -1:
+            logging.critical("Called send after close")
+            return
```
Comment (ercanucan) [inline endhost/scion_socket.py]: let's return 0 here?

File: endhost/scion_socket.py [hunk 4]
```diff
@@ -127,6 +127,9 @@ def send(self, msg):
...
         if msg is None or len(msg) == 0:
             return
```
Comment (ercanucan) [inline endhost/scion_socket.py]: ..and here as well (return 0)

File: endhost/scion_socket.py [hunk 5]
```diff
@@ -189,8 +195,22 @@ def getStats(self):
...
+            - shutdown()
+            - recv() until returns 0
+            - close()
+        :param how: Not implemented yet, placeholder for compatibility
+        :type how: int
+        """
+        self.libsock.SCIONShutdown(self.fd)
+
     def close(self):
         # TODO(ercanucan) : call the deleteSCIONSocket() once it is tested.
```
Comment (ercanucan) [inline endhost/scion_socket.py]: we can now remove this TODO now that we are calling deleteSCIONSocket()

File: endhost/scion_proxy.py (hunk 2 above)
Comment (aznair) [inline endhost/scion_proxy.py]: Yeah I just like it better when things are explicit, but agree it's not necessary

**Pull request merged**